model/diff/diff.go

```diff
@@ -0,0 +1,69 @@
+package diff
+
+import (
+	"github.com/aserto-dev/go-directory/pkg/derr"
+	"github.com/hashicorp/go-multierror"
+	"github.com/pkg/errors"
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination=mock_instances.go  -package=diff github.com/aserto-dev/azm/model/diff Instances
+
+type Diff struct {
+	Added   Changes
+	Removed Changes
+}
+
+type Changes struct {
+	Objects   []string
+	Relations map[string][]string
+}
+
+type Instances interface {
```

Comment:
Following a conversation with Gert, I think this interface needs to change.
In the edge-ds, it is very expensive to determine whether an instance of an object type or relation type exists. Doing so requires a full linear scan of the objects or relations. So calling it in a loop would be horribly slow.

It is more efficient to do a single scan and return all the types that do exist. 
For objects that's just a list of object types.
For relations it's a list of (object type, relation type, subject type, subject relation?) structures.

We should replace `ObjectExists` and `RelationsExist` with `ObjectTypes` and `RelationTypes` functions that will be called at most once (each).
If any relation type is removed in the diff, you'd call `RelationTypes` to get a list of all type combinations that appear in the relations table and search for the removed relations there.
If any object types are removed in the diff, you'd call `ObjectTypes()` in addition to `RelationTypes` to get the list of all object types that have instances and you can search for the removed types there.